feat: Add a toggle key for hide/show all the UI

sandrade-dcl opened the pull request:
# What? <!-- what is this PR? -->
We need a toggle key for hide/show all the UI in the screen. This will allow to make better screenshots or recordings to be shared by the community.

sandrade-dcl committed:
Create "ShowHideUIByTrigger" component

sandrade-dcl committed:
Merge branch 'master' into feat/Add-a-toggle-key-for-hide-show-all-the-UI

sandrade-dcl committed:
Apply "ShowHideUIByTrigger" component to the different UI prefabs

sandrade-dcl committed:
Merge branch 'master' into feat/Add-a-toggle-key-for-hide-show-all-the-UI

sandrade-dcl committed:
Merge branch 'master' into feat/Add-a-toggle-key-for-hide-show-all-the-UI

RominaMarchetti commented:
I've tried to use the other shortcuts while I the UI was invisible. 
Issue  1: after turning it on, when first I clicked M/Tab, instead of seeing the world plus its UI the map appeared. 
Issue 2: clicking enter while UI is off. The chat window opens and then U key doesn't show the UI because it is writing instead XD

Solution: block all key controls until player turns UI on back.

pravusjif commented on unity-client/Assets/Scripts/MainScripts/DCL/Controllers/HUD/Common/ShowHideUIByTrigger.cs:
```diff
@@ -0,0 +1,50 @@
+        toggleTrigger.OnTriggered += ToggleTrigger_OnTriggered;
+
+        canvasGroup = GetComponent<CanvasGroup>();
+        isVisible = canvasGroup.alpha > 0f;
```
I don't understand how this works for canvases created/configured **after** the triggering has happened.
for example:
1. I use the toggle key and then have no UI
2. I go into a new scene and the scene ui is invisible

I don't see a `ToggleTrigger_OnTriggered()` call inside this `Awake()`, so I don't understand how this works, can you explain that to me? thanks

It is also a bit confusing that the first `isVisible` value is based on the canvasGroup alpha, when that value can be different among canvases (in case any of our canvases already had a canvasGroup for example) and thus they could have a different isVisible initial value when it should be the same for all at all times.

sandrade-dcl replied:
- Related to why is it working for canvases created after the triggering happened: It is because almost all the prefabs involved in this feature have been prefabs that are loaded from the beggining: `Player`, `AvatarHUD`, `ExpressionsHUD`, `Taskbar`, etc.. But you are right, I didn't realize that could there exist other UI game objects created dinamically and then this flow wouldn't works for them. So I'm going to change the flow in order to take on account these cases. Good point, thanks!

- Related to base initially the `isVisible` variable on the alpha value of the CanvasGroup: You're right, I'm going to create a global boolean variable (Sriptable Object) to store the current status of the UI visibility and then my script will use it in order to know how to initialize itself. Also, this SO variable will be necessary for solving the previous point and for blocking all the input keys (the request from Romi). Thanks!

sandrade-dcl replied:
Done!

sandrade-dcl committed:
Several fixes

- Avoid to use input actions while player has the "AllUIVisibility" actived.
- Prepare de "Show/Hide UI" feature for UI's inside the parcels.

sandrade-dcl committed:
Merge branch 'master' into feat/Add-a-toggle-key-for-hide-show-all-the-UI

sandrade-dcl committed:
Remove commented code

sandrade-dcl committed:
Fix some conflicts with the ESC key

sandrade-dcl committed:
Block the "Show/Hide UI" action when settings or avatar editor are open

pravusjif commented on unity-client/Assets/Scripts/MainScripts/DCL/Controllers/HUD/Common/ShowHideUIByTrigger.cs:
```diff
@@ -1,44 +1,29 @@
-    private const string TOGGLE_UI_VISIBILITY_ASSET_NAME = "ToggleUIVisibility";
-
-    private InputAction_Trigger toggleTrigger;
+    //private InputAction_Trigger toggleTrigger;
```
Let's remove this commented code

sandrade-dcl replied:
Done!

sandrade-dcl committed:
Remove a commented code line

sandrade-dcl committed:
Block the "Show/Hide UI" action when the map is open

sandrade-dcl committed:
Merge branch 'master' into feat/Add-a-toggle-key-for-hide-show-all-the-UI

sandrade-dcl committed:
Merge branch 'master' into feat/Add-a-toggle-key-for-hide-show-all-the-UI
